src/main/scala/org/openhorizon/exchangeapi/route/user/Users.scala

```diff
@@ -214,6 +214,271 @@ trait Users extends JacksonSupport with AuthenticationSupport {
+                            .filter(_.password.isEmpty)
+                            .take(1)
+                            .joinLeft(UsersTQ.map(users => (users.organization, users.user, users.username)))
+                            .on(_.modifiedBy === _._2)
+                            .joinLeft(ApiKeysTQ)
+                            .on(_._1.user === _.user)
+                            .map(users =>
+                                   ((users._1._1.createdAt,
+                                     users._1._1.email,
+                                     users._1._1.identityProvider,
+                                     users._1._1.isHubAdmin,
+                                     users._1._1.isOrgAdmin,
+                                     users._1._1.modifiedAt,
+                                     users._1._1.modifiedBy,
+                                     users._1._1.organization,
+                                     None,
+                                     users._1._1.user,
+                                     users._1._1.username), users._1._2, users._2))))
+        } yield users.map(user => (user._1.mapTo[UserRow], user._2, user._3))
+        
+      complete({
+            db.run(getUserWithApiKeys.result.transactionally).map { result =>
+                  // logger.debug(s"GET /orgs/$organization/users/$username result size: " + result.size)
+
+                  if (result.nonEmpty) {
+                        val userResult = result.head
+                        val userRow = userResult._1
+                        val modifiedByInfo = userResult._2
+
+                        val apiKeyMetadataList = result.filter(_._3.isDefined).map { case (_, _, Some(apiKeyRow)) =>
+                              new ApiKeyMetadata(apiKeyRow, null)
+                        }.distinct
+
+                        val user = new User((userRow, modifiedByInfo), Some(apiKeyMetadataList))
+                        val userMap: Map[String, User] =
+                              Map(s"${userRow.organization}/${userRow.username}" -> user) // Ugly mapping, TODO: redesign response body
+
+                        (StatusCodes.OK, GetUsersResponse(userMap, 0))
+                  } else {
+                        (StatusCodes.NotFound, GetUsersResponse())
+                  }
+            }.recover {
+                  case t: org.postgresql.util.PSQLException =>
+                        ExchangePosgtresErrorHandling.ioProblemError(
+                              t, ExchMsg.translate("user.not.added", t.toString))
+
+                  case t =>
+                        (HttpCode.BAD_INPUT,
+                              ApiResponse(ApiRespType.BAD_INPUT, ExchMsg.translate("user.not.updated", t.toString)))
+            }
+      })
```

Comment:
This fragment is identical in `/orgs/$organization/users/apikey` and `/orgs/$organization/users/iamapikey` and almost the same as in `GET /orgs/$organization/users`. Please unify the code to avoid duplication.

Reply:
I do not insist, but I do not see a difference in the highlighted fragment between `/orgs/$organization/users/apikey` and `/orgs/$organization/users/iamapikey` - so I think we could just extract it to some function and call it from both routes. In general it is better to avoid code duplication as it is cheaper to maintain it in the longer run and less error prone. I do not suggest to combine the new routes with old ones - just to extract the overlapping fragments. If the idea is to keep only GET /orgs/$organization/users in the longer ru, we can leave it in the current shape, but the other two routes (apikey and iamapikey) can be easily combined into one route:
```
path("orgs" / Segment / "users" / ("apikey" | "iamapikey"))
```

Reply:
@wenyang-cao Changes look good. Thanks.